packages/teleterm/src/ui/utils/retryWithRelogin.ts

```diff
@@ -0,0 +1,71 @@
+import {} from 'teleterm/ui/services/workspacesService/documentsService';
+import { routing } from 'teleterm/ui/uri';
+import AppContext from 'teleterm/ui/appContext';
+
+export async function retryWithRelogin<T>(
+  appContext: AppContext,
+  originatingDocumentUri: string,
+  resourceUri: string,
```

Comment:
I ended up not needing `document.getResourceUri`, because I just pass this as an argument, so each place calling `retryWithRelogin` will be responsible for fetching its own resource URI.

Actually, `getResourceUri` would be hard to implement for `DocumentGateway`. What is the `resourceUri` for `DocumentGateway`? I'd say that it's `gatewayUri`, but this field is set only after the actual gateway is created. The gateway itself is kind of an ephemeral resource, it's deleted after you close the gateway or the app.

So in the case of `useDocumentGateway`, I actually pass `doc.targetUri` here.

I need `resourceUri` only to extract `rootClusterUri` from it. As I was writing this, I realized I could just accept `rootClusterUri` instead of `resourceUri`. However, it'd increase the complexity of using `retryWithRelogin` at the callsite, requiring each place to use something like `routing.ensureRootClusterUri`.

Fetching a relevant resource URI is much easier, usually we'll just grab it from a field in a doc.